data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java

```diff
@@ -0,0 +1,237 @@
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class KeyValueProcessorTests {
+    private KeyValueProcessor keyValueProcessor;
+    private PluginMetrics pluginMetrics = PluginMetrics.fromNames("kv", "pipeline");
+
+    @BeforeEach
+    void setup() {
+        keyValueProcessor = new KeyValueProcessor(pluginMetrics, new KeyValueProcessorConfig());
+    }
+
+    @Test
+    void testSingleKvToObjectKeyValueProcessor() {
+        final Map<String, Object> testData = new HashMap();
+        testData.put("message", "key1=value1");
+        final Record<Event> record = buildRecordWithEvent(testData);
+
+        final List<Record<Event>> editedRecords = (List<Record<Event>>)keyValueProcessor.doExecute(Collections.singletonList(record));
+
+        assertThat(editedRecords.size(), equalTo(1));
+        assertThat(editedRecords.get(0), notNullValue());
+        assertThat(editedRecords.get(0).getData().toJsonString(),
+                equalTo("{\"message\":\"key1=value1\",\"parsed_message\":{\"key1\":\"value1\"}}"));
+    }
+
+    @Test
+    void testMultipleKvToObjectKeyValueProcessor() {
+        final Map<String, Object> testData = new HashMap();
+        testData.put("message", "key1=value1&key2=value2");
+        final Record<Event> record = buildRecordWithEvent(testData);
+
+        final List<Record<Event>> editedRecords = (List<Record<Event>>)keyValueProcessor.doExecute(Collections.singletonList(record));
+
+        assertThat(editedRecords.size(), equalTo(1));
+        assertThat(editedRecords.get(0), notNullValue());
+        assertThat(editedRecords.get(0).getData().toJsonString(),
+                equalTo("{\"message\":\"key1=value1&key2=value2\",\"parsed_message\":{\"key1\":\"value1\",\"key2\":\"value2\"}}"));
+    }
+
+    @Test
+    void testSingleRegexFieldDelimiterKvToObjectKeyValueProcessor() {
+        keyValueProcessor = new KeyValueProcessor(pluginMetrics, getConfig("", null, "!_*!", "=", "", ""));
+
+        final Map<String, Object> testData = new HashMap();
+        testData.put("message", "key1=value1!_____!key2=value2");
+        final Record<Event> record = buildRecordWithEvent(testData);
+
+        final List<Record<Event>> editedRecords = (List<Record<Event>>)keyValueProcessor.doExecute(Collections.singletonList(record));
+
+        assertThat(editedRecords.size(), equalTo(1));
+        assertThat(editedRecords.get(0), notNullValue());
+        assertThat(editedRecords.get(0).getData().toJsonString(),
+            equalTo("{\"message\":\"key1=value1!_____!key2=value2\",\"parsed_message\":{\"key1\":\"value1\",\"key2\":\"value2\"}}"));
+    }
+
+    @Test
+    void testSingleRegexKvDelimiterKvToObjectKeyValueProcessor() {
+        //override setup
+        keyValueProcessor = new KeyValueProcessor(pluginMetrics, getConfig("", null, "&", ":\\+*:", "", ""));
```

Comment:
I recommend using the `createObjectUnderTest()` approach. You create an "object under test" in the `@BeforeEach`, but some tests use it and some don't. This can be error prone as the file changes over time.

I suggest that each test create its own object-under-test and there is no object-under-test field.